### ISSUE TITLE
Preparing for First Stable Release

### ISSUE THREAD

**Issue by TheLubab** [NEARBuilders/bos-workspace#48]
## Overview
This Epic tracks all the tasks and issues required to launch the first stable release of `bos-workspace`. Our goal is to ensure that the release is dev-friendly, well-tested, and well-documented.

#### Core functionalities
- [x] #31 
- [x] #22
- [ ] #24
- [ ] #19 
- [x] #30 
- [ ] #21
- [ ] #28 
- [ ] #26
- [x] #47 

#### Complete documentation
- [x] #46

#### Testing & Bugs
The code is missing tests, we need to improve the test coverage to ensure nothing breaks with new versions based on the core functionalities.

- [ ] #18

**Comment by elliotBraem** [comment]
I think we should prioritize the below for a first stable release:

- [ ] #31 
- [ ] #22  
- [ ] #47 

The rest are more wishlist items and will involve some explorations with other teams, such as #24 and #21, working with the near-query-api team -- and we'll probably want to work closer with [bos-cli-rs](https://github.com/bos-cli-rs/bos-cli-rs) for #26 and #19. #28 would be really cool, but could take its time; and honestly, #30 is just something to consider.

I think the target goal should be integrating with DevHub and Near Discovery, as this would help achieve a lot of synergy across the teams, and the only thing stopping this right now are the three I listed for priority. We need replacements for different environments, to match their existing replacements syntax, and make it as easy as installing the package and adding a bos.config.json.

It would be ideal to be able to update [devhub](https://github.com/NEAR-DevHub/neardevhub-bos) and [near-discovery](https://github.com/near/near-discovery-components) repositories with these steps:

- [ ] Add a bos.config.json that enables overrides for different environments, maybe points to the existing replacements jsons they have
- [ ] Configure the github workflows to deploy the app, build/src/*

Maybe something like this:

```
{
  appName: "app" // if app name is provided, then it will output to build/appName/src/*
  accounts: {
    deploy: "deploy.near",
    signer: "signer.near",
  },
  aliases: "./replacements.mainnet.json",
  overrides: {
    testnet: {
      accounts: {
        deploy: "deploy.testnet",
        signer: "signer.testnet",
      },
      aliases: "./replacements.testnet.json"
    }
    dev: {
      accounts: {
        deploy: "deploy.testnet",
        signer: "signer.testnet",
      },
      aliases: "./replacements.testnet.json"
    }
  }
}
```

Notice how [DevHub](https://github.com/NEAR-DevHub/neardevhub-bos/blob/main/replacements.mainnet.json) doesn't have a concept of an "appAccount"; it just has a list of replacements. REPL_DEVHUB is technically the appAccount; but this is helpful because this means replacements can be shared across projects without needed to convert an alias to "appAccount"

**Comment by TheLubab** [comment]
I totally agree with you about the priorities, it's better to release it with minimal features and keep expanding. 

In my current progress, I implemented the following:

- The parser, that replaces the custom imports, transpile ts... 
- The env in bos.config.json
- Simplify the apps directory
- The SocialDB data folder

To make it ready to use: 

- [ ] Recode/copy the dev and cli scripts from the old code.

To support migration from devhub and near-discovery:

- [ ] Change how aliases are called; from ${alias/NAME_OF_ALIAS} to ${NAME_OF_ALIAS}

Then we can start testing it.

**Comment by elliotBraem** [comment]
@sekaiking I also love the hot reload and local gateway; will this be going away?

**Comment by TheLubab** [comment]
No, the `bw dev` will stay the same. The only things missing will be the `bw upload` and `bw deploy` commands.

**Comment by elliotBraem** [comment]
This is the current outcome when downloading an account's widgets via bos-cli-rs (/src):

https://github.com/nearbuilders/hack.near ;)

**Comment by sethfork** [comment]
Any interest in exploring a more incremental pathway towards a stable release / v1.0.0? The current state of the branches is a little hard to follow. Some incremental releases could lower branch divergence and soften the migration hit
@elliotBraem @sekaiking @matiasberaldo

**Comment by elliotBraem** [comment]
Hope you've been well @sekaiking! 

Let me know what you think of the project thus far? I feel I've come to fully appreciate the thought and effort you put into this refactor. It has opened so many doors and possibilities. The app + workspace feature is really so nice. 

I think we're approaching a stable release; we still need `bw upload` which I haven't fully explored yet. And then soon, we will have a stable release and can close this issue!
